### ISSUE TITLE
BUG: Investigate Idle Timeout Behavior

### ISSUE THREAD

**Issue by cholly75** [ustaxcourt/ef-cms#8682]
**Describe the Bug**
We have had several reports that the behavior for the idle timeout functionality (the modal window that asks a user to confirm whether they are still active in their session) is not behaving as expected.

We have seen behavior in which:
1) The idle timeout modal does not appear on all screens after the 55 minute time limit (ex.: Docket Record screen, Pending Report)
2) After letting the next 5 minutes elapse after receiving the modal, the user is logged out - however the login screen presented to the user also shows various validation or confirmation messages (screenshots below):
3) After letting the next 5 minutes elapse after receiving the modal, the user is presented with a never-ending spinning wheel.  The solution is to just click the refresh button and they will get the log in screen.

<img width="623" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBNnROQkE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--bc6aa3257a539548309c692face53cfddce2d210/image.png" alt="image.png" />

<img width="529" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBNnhOQkE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--02e36608b4d1952218c0e92e5fdad0eaa4f4ee9e/image.png" alt="image.png" />

<img width="1531" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBN0xvQlE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--98d6691fa6f7df7ad746b52fbbd56e382274a86b/image.png" alt="image.png" />

**Business Impact/Reason for Severity**
Medium - confusing messaging/inconsistent behavior for users

**In which environment did you see this bug?**
PROD/TEST

**Who were you logged in as?**
Internal user and external user

**What were you doing when you discovered this bug? (Using the application, demoing, smoke tests, testing other functionality, etc.)**

**To Reproduce**
1. Navigate to the Docket Record view
2. Wait an hour

**Expected Behavior**
1. Modal presents to user after 55 minutes w/ no activity regardless of the actual page the user is on

**Actual Behavior**
1. Modal does not appear when looking at Docket Record for a case (among others).

**Screenshots**

<img width="360" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBN0pOQkE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--1a9ad2ed0ec526559091f713ad2250c52150ef41/image.png" alt="image.png" />

**Desktop (please complete the following information):**

**Smartphone (please complete the following information):**

**Cause of Bug, If Known**

**Process for Logging a Bug:**
* Complete the above information
* Add a severity tag (Critical, High Severity, Medium Severity or Low Severity). See below for priority definition. 

**Severity Definition:**
* Critical Defect
Blocks entire system's or module’s functionality
No workarounds available
Testing cannot proceed further without bug being fixed.

* High-severity Defect
Affects key functionality of an application
There's a workaround, but not obvious or easy
App behaves in a way that is strongly different from the one stated in the requirements

* Medium-severity Defect
A minor function does not behave in a way stated in the requirements.
Workaround is available and easy

* Low-severity Defect
Mostly related to an application’s UI
Doesn't need a workaround, because it doesn't impact functionality

## Definition of Ready for Bugs(Created 10-4-21)
Definition used: A failure or flaw in the system which produces an incorrect or undesired result that deviates from the expected result or behavior. (Note: Expected results are use cases that have been documented in past user stories as acceptance criteria and test cases, and do not include strange behavior unrelated to use cases.)

The following criteria must be met in order for the development team to begin work on the bug.

The bug must:
 - Be focused on solving a user problem
 - Contain data for all fields in the bug template, so the team can pick it up and begin working immediately

Process: If the unexpected results are new use cases that have been identified, but not yet built, new acceptance criteria and test cases should be captured in a new user story and prioritized by the product owner. 

If the Court is not able to reproduce the bug, add the “Unable to reproduce” tag. This will provide visibility into the type of support that may be needed by the Court. In the event that the Court cannot reproduce the bug, the Court will work with Flexion to communicate what type of troubleshooting help may be needed.

## Definition of Done (Updated 4-14-21)
**Product Owner**
 - [ ]  Bug fix has been validated in the Court's test environment

**Engineering**
 - [x] Automated test scripts have been written
 - [ ] Field level and page level validation errors (front-end and server-side) integrated and functioning
 - [ ] Verify that language for docket record for internal users and external users is identical
 - [ ] New screens have been added to pa11y scripts
 - [ ] All new functionality verified to work with keyboard and macOS voiceover https://www.apple.com/voiceover/info/guide/_1124.html 
 - [ ] READMEs, other appropriate docs and swagger/APIs fully updated
 - [ ] UI should be touch optimized and responsive for external only (functions on supported mobile devices and optimized for screen sizes as required)
 - [ ] Interactors should validate entities before calling persistence methods
 - [x] Code refactored for clarity and to remove any known technical debt
 - [ ] Deployed to the Court's `test` environment if prod-like data is required. Otherwise, deployed to any `experimental` environment for review.

**Comment by Mwindo** [comment]
I played around with this a bit by reducing the timeout locally.

So far, I am unable to reproduce issues 1 and 3: the modal appears on the Docket Record page, and I never received a never-ending spinning wheel. I will keep investigating.

That said, the fix for 2 should be as easy as adding `clearAlertsAction` to `gotoIdleLogoutSequence`.

**Comment by Mwindo** [comment]
Following a hunch, I was able to re-produce 1 and 3. It seems to have to do with having separate tabs open. For instance, one way to re-create the issue is to log in, go to a docket, and create a new order. This will open a separate tab. If you do nothing and wait until the timeout for the modal to appear runs out, the modal will only appear on the _first_ tab. However, once the subsequent timeout expires, regardless of what you are doing on the second tab, both tabs are logged out, and the second tab has a never-ending spinning wheel.

In other words, to reproduce:
- Log in as a judge (e.g.)
- Navigate to a docket record
- In the actions menu, click create a new order and navigate the modal. A new tab should open.
- Do nothing until the modal "Are you still there" should appear. It will not appear on this tab, but it should appear on the first tab.
- Now, whatever you do on the second tab is irrelevant: if you do not respond to the modal on the first tab, you will be logged out.

Of course, after writing this, I am having trouble re-creating again. But this process definitely produced the behavior a handful of times. [See Edit]

EDIT: If the first tab stays on Docket Record, I do not get the behavior. If I switch to a different tab (e.g., Drafts), then create the order, I do get the behavior.

EDIT 2: I'm seeing other potential weirdness as well. Example: I can log in, wait until I get the modal, copy the current URL, close the tab, wait for a long time, and paste back in the URL as if nothing happened (in other words, I wasn't logged out). This is "ok" in the sense that we have a token that expires in 1 day (in loginLamda.ts), but it is a little ... strange. This appears to be because we only track the idle time on the client side.

**Comment by Mwindo** [comment]
It's taken me a little while to fully grok the issue (and why it could happen enough to cause noticeable problems, but not enough to be a major pain point). I think what must be happening is the following:

- We do not set `state.lastIdleAction` until the user interacts with the app (where "interacts with" is defined [here](https://github.com/ustaxcourt/ef-cms/blob/67060668e4aad843544e7dd4e22f5630ebdb9539/web-client/src/views/IdleActivityMonitor.tsx#L25-L42)). This means that if a user logs in, refreshes, and does nothing, they will never receive the "Are you still there?" modal. The same goes for a user who opens a new DAWSON tab and does nothing. The reason for this is that we check idle logout behavior based on `state.lastIdleAction`; if it is `undefined`, the behavior is never triggered.
- Anytime the user interacts with the app, regardless of which tab, we broadcast a new `state.lastIdleAction` so that all tabs are in sync. Hooray!
- However, suppose the user who is working in DAWSON either 1) opens a new tab (maybe from "Create New Order") and then is pulled away for 55 mins or 2) is pulled away for 55 mins, forgets about their previous DAWSON tab, and opens to a new DAWSON tab. In this case, the original tab (where `state.lastIdleAction` has been set) will see that the modal should be displayed and start the countdown to automatic logout. The second tab, where `state.lastIdleAction` is still `undefined`, does not think the modal needs to be displayed, and the user sees nothing to indicate they will be logged out. At this point, no matter what the user does in this second tab, the first tab--which is now checking a `logoutAt` variable set when the modal appeared and no longer cares about the `state.lastIdleAction` values being broadcast from the user interaction in the second tab--will log them out of all tabs.


TL;DR: I think there is essentially a five-minute window--the time between the modal displaying and automatic logout--during which idle logout behavior can become out of sync so that one tab is aware of it but others are not.

**Comment by Mwindo** [comment]
A quick update/review, including notes on a couple of bugs that were not listed in the ticket description:

1) The user can refresh the login page and then never see the "Are you still there?" modal (and therefore never be automatically logged out) if they do not interact with the page.
- This is caused by us not updating `state.lastIdleAction` upon page load.
2) The user can sign in, sign out, wait 55 minutes, and then (rather humorously!) see the "Are you still there?" modal.
- This is caused by us not clearing `state.lastIdleAction` upon sign out. EDIT: Nope, this is a race condition artifact.
3) The user can see alert modals still present after being signed out.
- The is caused by us not always clearing alert modals upon sign out.
4) The user can end up with an infinite loading bar upon being signed out.
- This is caused by race conditions triggered by multiple calls to sign out after idle activity. (Every tab signs itself out but also broadcast to the other tabs to sign out, which can lead to actions being called out of order.)
5) Tabs can become "out of sync" within the 5 minutes between modal display and automatic sign out (see previous comment).
- This is caused by 1 and because we no longer check `state.lastIdleAction` once the modal has appeared, so no other tab can stop the automatic logout countdown.

I have tentative fixes for 1-4 in place and will await our decision about how we want to tackle 5.

**Comment by Mwindo** [comment]
I have a fix, with automated tests to boot (but I'll see what reviewers think of the slight hokey-ness it takes to make them work). (The hardest part--getting multi-tab testing to work--was the easy part. The easiest part--getting the tests that were working locally to work in CI--was the hardest part. I will double-check with other developers, but as far as I can tell, there were two unnecessary checks to prevent certain things from running in CI.)

I will deploy to test after the demo tomorrow and seek engineering reviews.

**Comment by ttlenard** [comment]
Testing Feedback @mwindo:

I was logged in as a Case Services Supervisor
I was on this page (notice the two different screen grabs, you can see the refresh errors in the console):
<img width="1559" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBMWMzQmc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--0209bc2ec727655776fbf03f67a392ad7cc19d3f/image.png" alt="image.png" />

<img width="1596" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBMWczQmc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--0b9acc1fd64977facb2bfaa7ae68351fb854f913/image.png" alt="image.png" />

I never received the "are you still there?" pop-up
When I clicked on the Messages tab, I got the never-ending spinning wheel, and had to refresh the page to get out of it and get back to the login screen.
<img width="1562" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBMXMzQmc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--9e9b9de1785ae612d180687fa615b8b1adb75e6f/image.png" alt="image.png" />

<img width="1541" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBMXczQmc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--dbb3cf24af49f4c391e86bde161d811ae5cd2416/image.png" alt="image.png" />

**Comment by Mwindo** [comment]
I have not been able to re-create the exact behavior as above. However, I have noticed similar issues when a new deploy to test happens while a user is still signed into test (both the "Close and Refresh" and "Are you still there?" modals appear, along with the loading screen, in a chaotic fashion). I will update when I have more information.

**Comment by Mwindo** [comment]
My best guess as to what happened given a recent test:

- Tenille's browser was making intermittent requests to refresh (as expected).
- At some point, the new deploy to test occurred. At this point, we switched from green to blue, and the refresh requests failed because they were being directed to the wrong endpoint. (As seen in Tenille's first screenshot.)
- For some reason, the "Close and Refresh" modal and "Are you still there?" modal interfered with each other, and neither was shown. This allowed Tenille to interact with the screen as if everything was "ok."
- But everything was *not* "ok"--Tenille needed to refresh to direct her requests to the correct endpoint. When she tried to navigate elsewhere in the app, she got a CORS error (as seen in Tenille's third screenshot).

I am still looking into why the modal never appeared; I suspect a race condition. Nonetheless, by pretending the "Close and Refresh" modal in the below screen shot *wasn't* there (e.g., doing some JavaScript stuff to "click" to go to another page despite the modal), I have simulated the above steps and gotten the same behavior as Tenille reported above:

<img width="1237" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBNUJEQmc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--a1fcd98a4618c8548d2b0aae7e4285d8054be5f7/Screenshot%202024-08-08%20at%201.48.23%E2%80%AFPM.png" alt="Screenshot 2024-08-08 at 1.48.23 PM.png" />
:

<img width="1237" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBNDVEQmc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--b66f5915760ef1ce543ff206f65da70c7aea8436/Screenshot%202024-08-08%20at%202.15.05%E2%80%AFPM.png" alt="Screenshot 2024-08-08 at 2.15.05 PM.png" />

**Comment by Mwindo** [comment]
I am still trying to determine the _exact_ mechanism for what I suspect happened (mainly, why is the URL in Tenille's second screen shot the same when I expect it to have been routed elsewhere), but I think the above outline I commented is the gist, except I now suspect the modals _did_ appear (i.e., no race condition) but were cleared.

I suspect the "Close and Refresh" modal showed up after the first failed request to the stale endpoint; later, the "Are you still there?" modal appeared at the designated time; 5 minutes later, the sign out sequence attempted to sign Tenille out; however, it (specifically, the `deleteAuthCookieAction`) failed because the endpoint was stale (this is the failed network request to `auth/login`, the first request in Tenille's second screen shot), the modals were cleared, and there was no sign of any issue. When Tenille attempted to go elsewhere, the same end behavior I outlined above occurred.

**Comment by Mwindo** [comment]
After a brief discussion with the team, here are the following points I will address next:

1. Because we do not currently broadcast "DAWSON has been updated" to other tabs, I believe we can get mismatches across tabs (similar to the mismatches that were occurring with the "Are you still there?" modal). This isn't a big deal--it essentially amounts to one tab showing the message to "Close and Refresh" before another--but it would be better to be consistent.
2. Currently, we can end up making sign-out requests to a stale endpoint, which leads to problematic behavior like that documented by Tenille above.
3. In general, we should avoid making requests that we "know" will fail.

To address these points, we can send/handle a broadcast message upon receiving a "X_FORCE_REFRESH" header in a request response that will ensure the modal is shown across tabs (which will address 1) and set some state, e.g.,  `dawsonHasUpdated`. (EDIT: I decided to call this `clientNeedsToRefresh` instead.) We can check this state before making certain requests, e.g., trying to sign out the user automatically (the source of issue 2) or making a refresh token request.

**Comment by ttlenard** [comment]
@Mwindo Some testing feedback.

1. Log in as a Judge or Chambers user
2. Navigate to a case
3. Open the console and paste your snippet to lessen the timeout time (I set it to 10 seconds)
4. Click on create, and then select Order
5. This will open a new tab
6. Start to draft an order on tab 2
7. Be idle for 10 seconds; you will not get the "are you still there?" message, but it WILL display on tab 1 (unbeknownst to the user)
8. After 10 more seconds, you will be logged out (since you didn't click yes that you were still there on tab 1)

Can we get the "Are you still there?" modal on the additional tab(s) that the user has open?
Chris Holly was also ok with adjusting the idle timeout time to more than 55 minutes...adjust to 4 hours? Leave it as is?

**Comment by Mwindo** [comment]
The issue above was caused because we were not running the script in both tabs. (So the first tab was waiting 10 seconds, but the new tab was waiting 55 minutes.) It is probably even _better_ to coordinate the display of the modal  across tabs directly (we currently coordinate this across tabs indirectly based on the last user idle time), but in practice the two should never become decoupled since the idle wait times are constants (that we were fudging with in the console).

As for the idle wait time constant itself being updated: I don't think there is any reason not to do so for security purposes, at least currently. At present, the idle time doesn't apply when windows are closed, so a user could log in, close their DAWSON window, and then reopen a DAWSON window hours later and still be logged in. The "real" security is enforced by the 24-hour token.
